### ISSUE TITLE
`histogram_quantile` get incorrect result when samples miss.

### ISSUE THREAD

**Issue by eddietanshuo** [VictoriaMetrics/VictoriaMetrics#2819]
**Describe the bug**
Hello @valyala @YuriKravetc, we find a new issue in our product env.
We get 2 datasource,  the left one is our prometheus native storage and the right one is our victoriametrics storage which is remote written by the prometheus.
![image](https://user-images.githubusercontent.com/103412360/177039962-e8b3de9a-3225-4a0b-aa18-3a4ac31e42c8.png)

As the above image shows, victoriametrics sometimes has very large value that could result in wrong alert.

The expression is like:
```
histogram_quantile(0.99, sum(rate(etcd_disk_wal_fsync_duration_seconds_bucket{instance=~"xxxxx",job=~"xxxx"}[5m])) by (instance, le))
```

I expand the result by showing all buckets:

  le| left datasource | right datasource
-- | -- | --
0.001 | 0.00351 | 0.00333
0.002 | 0.0351 | 0.0417
0.004 | 0.0386 | 0.0367
0.008 | 0.0421 | 0.0400
0.016 | 0.0491 | 0.0467
0.032 | 0.0491 | 0.0467
0.064 | 0.0491 | 0.0467
0.128 | 0.0491 | 0.0467
0.256 | 0.0491 | 0.0467
0.512 | 0.0491 | 0.0467
1.024 | 0.0491 | 0.0467
2.048 | 0.0491 | 0.0467
4.096 | 0.0491 | 0.0467
8.192 | 0.0491 | 0.0583
INF | 0.0491 | 0.0467

It's very strange that victoriametrics has a larger value with a smaller `le`, which should be the reason of incorrect result.
![image](https://user-images.githubusercontent.com/103412360/177040213-b1585cf4-580e-4e20-aacc-adf152e797db.png)

**Cause**
I plot a new graph for the rate of `le=4.096` and the rate of `le=8.192` and find the problem: `le=8.192` which is the green line in the below picture has a unexpected big value during the problem time range. However, `le=8.192` and `le=4.096` should have the same rate since they have the same value all over time.
![image](https://user-images.githubusercontent.com/103412360/177040372-022eec40-b8e5-4d4d-bb55-e80a0f054f1c.png)I 

I add log for all rollup window's calculation and find the diff: 
![image](https://user-images.githubusercontent.com/103412360/177040559-2671de3b-7615-4849-bc02-e72a89c06f57.png)

The left file is the log for the rate calculation of `le=8.192`, it's tEnd stay the same for one more time period. Since the rate is calculated by `(vEnd - vPrev) / (tEnd - tPrev)`, so it's rate value is larger than what it should be.

For more detail, I print the raw sample timestamp for `le=8.192` and `le=4.096`, there are two points missed with timeseries `le=8.192` at the problem time range: 
![image](https://user-images.githubusercontent.com/103412360/177040759-125a6645-d844-41ec-bebb-d5eab6da3ac5.png)

Since victoriametrics will always use the last timestamp in a window rollup value calculation, the tEnd will remain the same in the next window if sample miss:

![image](https://user-images.githubusercontent.com/103412360/177040928-34e14678-449d-4ff4-92b6-37d6c1952d2a.png)

**What we expect**

Since sample miss is inevitable in remote write,  we hope `histogram_quantile` calculation should not  increase to a very large value in the case. For example: when calculate `histogram_quantile`, if we keep the value of `le=8.192` always larger than or equal to the value of `le=4.096` the graph of `histogram_quantile` will be much more reasonable.

**Comment by valyala** [comment]
@eddietanshuo , thanks for the outstanding analysis of the issue. I'm going to modify the code for `histogram_quantile()`, so it makes sure that values for `le` buckets don't increase with the increased `le`. If this is the case then override the value for smaller `le` with the closest value for higher `le`. For example, the value for `le="8.192"` above would be substituted with the value for `le="+Inf"`, e.g. 0.0467, while the value for `le="0.002"` would be substituted with the value for `le="0.004"`, e.g. 0.0367.

**Comment by valyala** [comment]
The issue should be fixed in the following commits:

- single-node VictoriaMetrics - f3ece83e67bb6636b042cc1ebad1e7d71e4534cf
- VictoriaMetrics cluster - 85166705827310062c8cae1b2b1e37bc6ac09530

These commits will be included in the next release.

@eddietanshuo , could you build VictoriaMetrics from these commits and verify whether this fixes the issue? See [build instructions for single-node VictoriaMetrics](https://docs.victoriametrics.com/#how-to-build-from-sources) and [build instructions for cluster version of VictoriaMetrics](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#building-from-sources)

**Comment by eddietanshuo** [comment]
Thanks!! We expect the new version release.

**Comment by lemonup** [comment]
we have the same issue with wrong rates e.g. `le=0.5` and `le=1` (`le=0.5` has greater value than `le=1`) with 
`histogram_quantile(0.9999 ...)` on release v1.79.0
it seems to have the same root
is it possible to fix it?
I could provide more info if needed

**Comment by mozedz** [comment]
I have faced the same issue on v1.80.0 as well, after downgrading to v1.78.0, the issue was gone.

**Comment by valyala** [comment]
> we have the same issue with wrong rates e.g. le=0.5 and le=1 (le=0.5 has greater value than le=1) with
histogram_quantile(0.9999 ...) on release v1.79.0
it seems to have the same root
is it possible to fix it?
I could provide more info if needed

It would be great if you upgrade to the latest available release - [v1.81.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases) now - and share the `<query>` together with the results of the `<query>` from `histogram_quantile(0.9999, <query>)` alongside the unexpected results of `histogram_quantile(0.9999, ...)` together with the expected results.

**Comment by mozedz** [comment]
![prom](https://user-images.githubusercontent.com/8280546/190061744-c3b4bef8-fd41-4c00-8376-21c03e24374c.png)

Hi @valyala , I have tested with 2 versions: the left is `v1.81.2-cluster` the right is `v1.78.1-cluster`, seems both worked as expected. Confirmed the issue is gone.

**Comment by valyala** [comment]
> Do you need me to test again with v1.80.x ?

No, this isn't needed, since v1.80.x isn't an LTS (long-time support) release unlike [v1.79.x](https://docs.victoriametrics.com/CHANGELOG.html#v1790) releases.

**Comment by cheqrouni** [comment]
Hello,
We are experiencing the same problem with CoreDNS request latency.

Please find below a screenshot of our dashboard with Prometheus and Victoriametrics data sources:

CoreDNS request latency with Prometheus :

![image](https://user-images.githubusercontent.com/7829065/213206510-966d6533-bd10-471a-a9aa-e1f968b501ed.png)

CoreDNS request latency with Victoriametrics :

![image](https://user-images.githubusercontent.com/7829065/213206668-85480f6f-a307-49be-946c-9fd4048afa82.png)

Any update for a LTS version?

Thank you!

**Comment by Whisper40** [comment]
Hello @valyala, any update on this issue ? Does it has been fixed in the LTS version ?

**Comment by hagen1778** [comment]
@Whisper40 @cheqrouni have you tried resetting the caches before doing tests?

**Comment by hagen1778** [comment]
The logic for calculating histogram_quantile for inconsistent buckets has been changed in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6547
The change should provide more reliable results in case if histogram buckets were partially updated.

**Comment by valyala** [comment]
FYI, the changed logic is available in [VictoriaMetrics v1.93.16 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.93.16).

**Comment by valyala** [comment]
FYI, the changed logic is available in [VictoriaMetrics v1.97.6 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.97.6).

**Comment by valyala** [comment]
FYI, the changed logic is available in [VictoriaMetrics v1.102.0 release](https://docs.victoriametrics.com/changelog/).

**Comment by panpan0000** [comment]
why this issue still open while 1.93.16+ has fixed it ?

**Comment by makasim** [comment]
Closing this as completed. Feel free to reopen it if anything else comes up— we're happy to help!
